### PR TITLE
Issue 2020 change labels on message details screen

### PR DESCRIPTION
**Pull request by DenBond7**
This PR added the ability to change message labels

issue #2020

----------------------------------

**Tests** _(delete all except exactly one)_:
- Tests will be added later (issue #2464)

--------------------------------

### To be filled by reviewers

I have reviewed that this PR... _(tick whichever items you personally focused on during this review)_:
- [x] addresses the issue it closes (if any)
- [x] code is readable and understandable
- [x] is accompanied with tests, or tests are not needed
- [x] is free of vulnerabilities

### CODE REVIEW

File: FlowCrypt/src/main/java/com/flowcrypt/email/jetpack/viewmodel/GmailLabelsViewModel.kt [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+  private val controlledRunnerForChangingLabels = ControlledRunner<Result<Boolean>>()
+  private val changeLabelsMutableStateFlow: MutableStateFlow<Result<Boolean?>> =
+    MutableStateFlow(Result.none())
+  val changeLabelsStateFlow: StateFlow<Result<Boolean?>> =
+    changeLabelsMutableStateFlow.asStateFlow()
+
+  @OptIn(ExperimentalCoroutinesApi::class)
+  val labelsInfoFlow: Flow<List<LabelWithChoice>> =
+    activeAccountLiveData.asFlow().mapLatest { account ->
+      if (account?.isGoogleSignInAccount == true) {
```
Comment (ioanmo226) [inline FlowCrypt/src/main/java/com/flowcrypt/email/jetpack/viewmodel/GmailLabelsViewModel.kt:42]: Might be better with takeIf and let

```kt
accountEntity?.takeIf { it.isGoogleSignInAccount }?.let { currentAccount ->
...
} ?: emptyList()
```

File: FlowCrypt/src/main/java/com/flowcrypt/email/database/dao/MessageDao.kt [hunk 2]
```diff
@@ -506,6 +506,25 @@ abstract class MessageDao : BaseDao<MessageEntity> {
...
+    withContext(Dispatchers.IO) {
+      if (email == null || label == null) {
+        return@withContext
+      }
+
+      val messagesToBeUpdated = mutableListOf<MessageEntity>()
+      val entities = getMsgsByUIDs(email, label, labelsToBeUpdatedMap.keys.toList())
+      for (entity in entities) {
+        messagesToBeUpdated.add(entity.copy(labelIds = labelsToBeUpdatedMap[entity.uid]))
+      }
```
Comment (ioanmo226) [inline FlowCrypt/src/main/java/com/flowcrypt/email/database/dao/MessageDao.kt:523]: Better to add map for this case. We don't need to declear above messagesToBeUpdated in this case.

```kt
val updatedMessages = entities.map { entity ->
  entity.copy(labelIds = labelsToUpdate[entity.uid])
}
```

**Review by ioanmo226 — COMMENTED** [review]
Below small recommendations.
Also when I try to pull from this branch and run app below error occurs.
I think you need to update migration version number?

>  
> java.lang.RuntimeException: Exception while computing database live data.
> 	at androidx.room.RoomTrackingLiveData.refreshRunnable$lambda$0(RoomTrackingLiveData.kt:74)
> 	at androidx.room.RoomTrackingLiveData.$r8$lambda$PhMGW5zFk4QWazERd2lfEeLZqW0(Unknown Source:0)
> 	at androidx.room.RoomTrackingLiveData$$ExternalSyntheticLambda0.run(Unknown Source:2)
> 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
> 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
> 	at java.lang.Thread.run(Thread.java:920)
> Caused by: java.lang.IllegalStateException: Room cannot verify the data integrity. Looks like you've changed schema but forgot to update the version number. You can simply fix this by increasing the version number. Expected identity hash: c5b1d7d328ad58717e08bbf227c807b0, found: df758f7866f01e7b79bf36fe927486ae
>

**Comment by ioanmo226** [comment]
I think better to update migration version too for development build?
Otherwise developers have to delete and re-run app?

**Comment by DenBond7** [comment]
Sorry for the delayed reply.

> Otherwise developers have to delete and re-run app?

You are right

> I think better to update migration version for development build too?

Let me add more details 

Please look at this image

image

At step **1**
image
we have the latest published version that is visible to users on `Google Play`. It's the release commit. This commit should have a tag(1.4.6 for this one). Thanks to this we can easily switch between `Google Play` snapshots and test what exactly users have.

image

At step **2** I've made the database migration `38 -> 39`
image

At step **3** I've made the database migration `39 -> 40`
image

It looks like I've made a mistake and made one more migration during releases. Because I follow the following rule - one migration between public releases.

If I make any database changes in a current milestone before the next release I use the following steps to check if all works well:

1. Switch to a commit that relates to the latest public release(for example `1.4.6` tag)
2. Build and install the app. Set up at least one account.
3. Switch to the last commit
4. Build and install the app(Should be read as update existing)
5. Check that there are no migration issues. It's easy to understand. any migration issue will not allow to use the app.

All of the above means I don't care about proper migration during the development process in the active milestone. I just need to be sure that for example `1.4.6` will be upgraded to `1.4.7` without issues.

### Between releases strategy
**pros**

- fast development

**cons**

- in most cases need to reinstall the app(Ioan's situation)


### Every PR strategy
**pros**

- Easy testing. Don't need to re-install the app
- Encapsulated logic. The app at any next commit should work without database migration issues.

**cons**

- more time for development. Database migration is not an easy action. Need additional time for every migration and additional time for testing.
- The code will look more complicated. More migration -> more places to make a mistake. due to this, I try to do migration as a last resort.

IMHO, it's better to re-install the app and spend up to 10 minutes to setup it up instead of spending hours to add a proper migration. A few years ago, after a little discussion with Tom, we decided to use `Between releases strategy`. But time is going and maybe need to relook this(as Ioan proposed).

@tomholub, @sosnovsky what do you think?

**Comment by sosnovsky** [comment]
Having migration in the same PR where it should be performed seems to be a better choice for me, as 1 migration per release looks more complex. And I think overall it's more common implementation for migrations, I don't remember another project where several migrations were grouped into one, to make 1 migration per release.

@DenBond7 for `Between releases strategy` pros you mentioned 'fast development', but why it's more complicated to have several smaller migrations per release, than 1 big migration?

**Comment by DenBond7** [comment]
@sosnovsky thank you for your feedback
> for Between releases strategy pros you mentioned 'fast development', but why it's more complicated to have several smaller migrations per release, than 1 big migration?

Because 1 migration -> 1 testing process, 2 migrations -> 2 testing processes, and so on. 

Usually, one release doesn't contain a lot of database changes. (like the current one). And instead of having 

```kotlin
@VisibleForTesting
    val MIGRATION_38_39 = object : FlowCryptMigration(38, 39) {
      override fun doMigration(database: SupportSQLiteDatabase) {
        //ref https://github.com/FlowCrypt/flowcrypt-android/issues/2448
        database.execSQL("ALTER TABLE labels ADD COLUMN label_color TEXT DEFAULT NULL;")
        database.execSQL("ALTER TABLE labels ADD COLUMN text_color TEXT DEFAULT NULL;")
      }
    }

    @VisibleForTesting
    val MIGRATION_39_40 = object : FlowCryptMigration(39, 40) {
      override fun doMigration(database: SupportSQLiteDatabase) {
        //ref https://github.com/FlowCrypt/flowcrypt-android/issues/2020

        //need to delete all messages where folder = 'INBOX' to reload messages and save labelIds
        database.execSQL("DELETE FROM messages WHERE folder = 'INBOX'")
        database.execSQL("ALTER TABLE messages ADD COLUMN `label_ids` TEXT DEFAULT NULL;")
        database.execSQL("ALTER TABLE labels ADD COLUMN `label_list_visibility` TEXT NOT NULL DEFAULT 'labelShow';")
      }
    }
```

I can have just 

```kotlin
@VisibleForTesting
    val MIGRATION_38_39 = object : FlowCryptMigration(38, 39) {
      override fun doMigration(database: SupportSQLiteDatabase) {
        //ref https://github.com/FlowCrypt/flowcrypt-android/issues/2448
        database.execSQL("ALTER TABLE labels ADD COLUMN label_color TEXT DEFAULT NULL;")
        database.execSQL("ALTER TABLE labels ADD COLUMN text_color TEXT DEFAULT NULL;")
        //ref https://github.com/FlowCrypt/flowcrypt-android/issues/2020
        //need to delete all messages where folder = 'INBOX' to reload messages and save labelIds
        database.execSQL("DELETE FROM messages WHERE folder = 'INBOX'")
        database.execSQL("ALTER TABLE messages ADD COLUMN `label_ids` TEXT DEFAULT NULL;")
        
        database.execSQL("ALTER TABLE labels ADD COLUMN `label_list_visibility` TEXT NOT NULL DEFAULT 'labelShow';")
      }
    }
```

But after Ioan's question and your reply, I'm thinking maybe having to have `Every PR strategy` is a better way. If use it we will have an always-ready master branch(It means the master branch by default contains available for release commits). I mean I don't need to approve a new release and anybody from the team can make a new release without scaring to break something. 

previously, Tom asked me if the code was ready to be published. But it was a bit far when I was a single developer in the project and we didn't use PRs.

**Comment by sosnovsky** [comment]
> Because 1 migration -> 1 testing process, 2 migrations -> 2 testing processes, and so on.

Maybe I'm missing something, but testing process looks same for me:
- there is PR 1 with migration `MIGRATION_38_39`
- you test it by installing build from previous release and then updating it
- then PR 1 is merged to `master`
- later you have another PR 2 with migration `MIGRATION_39_40`
- you test it the same way - by installing build from previous release and then updating it

With your current approach testing looks the same, but you're not adding another migration `MIGRATION_39_40`, but add new queries to the existing migration `MIGRATION_38_39`

**Comment by sosnovsky** [comment]
> But after Ioan's question and your reply, I'm thinking maybe having to have Every PR strategy is a better way. If use it we will have an always-ready master branch(It means the master branch by default contains available for release commits). I mean I don't need to approve a new release and anybody from the team can make a new release without scaring to break something.

It's a great advantage too, as it's expected for `master` branch to always have ready to release code (we use this approach in iOS and browser extension repos)

**Comment by DenBond7** [comment]
> Maybe I'm missing something, but testing process looks same for me: ...
... there is PR 1 with migration

You are right. But some time ago we didn't use PR. We used to create a branch for a milestone, I worked in that branch and Tom just merged that branch to the master before release. In those days my logic made sense. But now... it looks weird.

**Comment by DenBond7** [comment]
From now, I will use `Every PR strategy`

**Comment by sosnovsky** [comment]
> You are right. But some time ago we didn't use PR. We used to create a branch for a milestone, I worked in that branch and Tom just merged that branch to the master before release. In those days my logic made sense. But now... it looks weird.

Ah, I see, then it was easier approach, but with PRs having separate migrations is more convenient.
Glad that we've found solution for it :)

Thanks Ioan for noticing 👍

**Comment by DenBond7** [comment]
@ioanmo226 I've added changes. Please try.

**Comment by ioanmo226** [comment]
@DenBond7 Just checked, it's running smoothly now.
But, found a small glitch. If you open an email in the inbox, remove the `Inbox` label, and then go back, the email still shows up there.
It should disappear I think.

**Comment by DenBond7** [comment]
> @DenBond7 Just checked, it's running smoothly now. But, found a small glitch. If you open an email in the inbox, remove the `Inbox` label, and then go back, the email still shows up there. It should disappear I think.

You are right. It's a good one, I've missed it. The message should disappear. For now, you can do pull-to-refresh. It will delete the message from the screen. But it should work automatically. I will fix it. 

By the way, it relates to all labels, not only INBOX. If you open for example label = "Invoices" from the side menu, open some message, and remove `Invoices` label the message should disappear from the list too.

**Comment by ioanmo226** [comment]
Yeah!

**Comment by DenBond7** [comment]
@ioanmo226 I've added the requested changes. Please have a look

**Review by ioanmo226 — APPROVED** [review]
:+1

**Pull request merged**